Civi/Core/Url.php

```diff
@@ -0,0 +1,690 @@
+   * with their URL-encoded values.
+   *
+   * Example:
+   *   Civi::url('frontend://civicrm/greeter?cid=[contact]&msg=[message]')
+   *     ->addVars(['contact' => 123, 'message' => 'Hello to you & you & you!');
+   *
+   * @param string[] $vars
+   * @return $this
+   */
+  public function addVars(array $vars): Url {
+    $this->vars = $vars + ($this->vars ?: []);
+    return $this;
+  }
+
+  /**
+   * @return callable|null
+   */
+  public function getVarsCallback(): ?callable {
+    return $this->varsCallback;
+  }
+
+  /**
+   * Configure dynamic lookup for variables.
+   *
+   * @param callable|null $varsCallback
+   *   Function(string $varName): ?string
+   *   Determine the string-value of the variable. (May be ''.)
+   *   If the variable is unavailable, return NULL.
+   * @return $this
+   */
+  public function setVarsCallback(?callable $varsCallback) {
+    $this->varsCallback = $varsCallback;
+    return $this;
+  }
+
+  /**
+   * Apply a series of flags using short-hand notation.
+   *
+   * @param string $flags
+   *   List of flag-letters, such as (a)bsolute or (r)elative
+   *   For a full list, see Civi::url().
+   * @see Civi::url()
+   * @return $this
+   */
+  public function useFlags(string $flags): Url {
+    $len = strlen($flags);
+    for ($i = 0; $i < $len; $i++) {
+      switch ($flags[$i]) {
+        // (a)bsolute url
+        case 'a':
+          $this->preferFormat = 'absolute';
+          break;
+
+        // (r)elative url
+        case 'r':
+          $this->preferFormat = 'relative';
+          break;
+
+        // (h)tml encoding
+        case 'h':
+          $this->htmlEscape = TRUE;
+          break;
+
+        // (t)ext encoding (canonical URL form)
+        case 't':
+          $this->htmlEscape = FALSE;
+          break;
+
+        // (s)sl
+        case 's';
+          $this->ssl = TRUE;
+          break;
+
+        // (c)ache code for resources
+        case 'c':
+          $this->cacheCode = TRUE;
+          break;
+      }
+    }
+    return $this;
+  }
+
+  /**
+   * Render the final URL as a string.
+   *
+   * @return string
+   */
+  public function __toString(): string {
+    $userSystem = \CRM_Core_Config::singleton()->userSystem;
+    $preferFormat = $this->getPreferFormat() ?: static::detectFormat();
+    $scheme = $this->getScheme();
+
+    if ($scheme === NULL || $scheme === 'current') {
+      $scheme = static::detectScheme();
+    }
+
+    if ($scheme === 'default') {
+      $scheme = \CRM_Core_Menu::isPublicRoute($this->getPath()) ? 'frontend' : 'backend';
+    }
+
+    switch ($scheme) {
+      case 'frontend':
+      case 'service':
+        $result = $userSystem->url($this->getPath(), $this->getQuery(), $preferFormat === 'absolute', $this->composeFragment(), TRUE, FALSE, FALSE);
+        break;
+
+      case 'backend':
+        $result = $userSystem->url($this->getPath(), $this->getQuery(), $preferFormat === 'absolute', $this->composeFragment(), FALSE, TRUE, FALSE);
+        break;
```

Comment:
Am i right in thinking that routing to `$userSystem->url()` here provides an opportunity (sometime in the future) to replace it with something less problematic?

Reply:
@christianwach I do like the idea of `$userSystem` having some more influence/visibility on the $scheme. For example, I don't think it's great for `frontend://` and `service://` to be conjoined like that. I'll add on some ideas for that...

Reply:
@christianwach So does 59af62f8b3d280cf3080d78be3d9a86c28ab2749 seem better/worse/same? It delegates to `$userSystem->getRouteUrl()` instead of `$userSystem->url()`. The main differences are:

* `UF::getRouteUrl()` receives the name (`frontend`, `backend`, `service`) instead of multiple bools. It can potentially receive other/custom schemes if the UF has other environments.
* `UF::getRouteUrl()` doesn't have to deal with the fragment or absolute/relative flag. (These seem amenable to generic handling.)

Another thing the UF might influence is the interpretation of `current://` (ie `detectScheme()`). With the current PR, the UF-integration could [set `civicrm_url_defaults`](https://github.com/civicrm/cv/pull/171/files)  before invoking the page.

Reply:
> So does https://github.com/civicrm/civicrm-core/commit/59af62f8b3d280cf3080d78be3d9a86c28ab2749 seem better/worse/same?

@totten Better in that the inscrutable boolean params are somewhat hidden. Nice one :)